core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java

```diff
@@ -0,0 +1,113 @@
+import static io.netty.handler.codec.haproxy.HAProxyProtocolVersion.V2;
+import static io.netty.handler.codec.haproxy.HAProxyProxiedProtocol.TCP4;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.proxy.HAProxyConfig;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.haproxy.HAProxyMessage;
+import io.netty.handler.codec.haproxy.HAProxyMessageEncoder;
+import io.netty.handler.proxy.ProxyConnectException;
+
+class HAProxyHandler extends ChannelOutboundHandlerAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(HAProxyHandler.class);
+
+    private final HAProxyConfig haProxyConfig;
+
+    HAProxyHandler(HAProxyConfig haProxyConfig) {
+        this.haProxyConfig = haProxyConfig;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().addBefore(ctx.name(), null, HAProxyMessageEncoder.INSTANCE);
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
+                        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        promise.addListener(f -> {
+            if (!f.isSuccess()) {
+                return;
+            }
+            try {
+                ctx.write(createMessage(haProxyConfig, ctx.channel())).addListener(f0 -> {
+                    if (f0.isSuccess()) {
+                        ctx.pipeline().remove(HAProxyMessageEncoder.INSTANCE);
+                    } else {
+                        ctx.fireExceptionCaught(new ProxyConnectException(f0.cause()));
+                        ctx.close();
+                    }
+                });
+            } catch (Exception e) {
+                ctx.channel().eventLoop().execute(
+                        () -> ctx.pipeline().fireUserEventTriggered(new ProxyConnectException(e)));
+                ctx.close();
+            } finally {
+                ctx.pipeline().remove(this);
+            }
+        });
+        super.connect(ctx, remoteAddress, localAddress, promise);
+    }
+
+    private static HAProxyMessage createMessage(HAProxyConfig haProxyConfig,
+                                                Channel channel) throws ProxyConnectException {
+        final InetSocketAddress srcSocketAddress =
+                haProxyConfig.sourceAddress() != null ? haProxyConfig.sourceAddress()
+                                                      : (InetSocketAddress) channel.localAddress();
+        final InetSocketAddress destSockAddress = haProxyConfig.proxyAddress();
```

Comment:
nit: consistency? socket or sock
Or we could just `srdAddress` and `destAddress`